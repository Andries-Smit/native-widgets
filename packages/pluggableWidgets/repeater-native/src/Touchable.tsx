import { createElement, FunctionComponent, Fragment } from "react";
import {
    TouchableNativeFeedback,
    TouchableHighlight,
    Platform,
    ViewStyle,
    LayoutChangeEvent,
    TouchableOpacity,
    TouchableNativeFeedbackProps,
    TouchableHighlightProps,
    View
} from "react-native";

const DEFAULT_RIPPLE_COLOR = "rgba(0, 0, 0, 0.2)";
const isAndroid = Platform.OS === "android";

export interface TouchableProps extends TouchableNativeFeedbackProps, TouchableHighlightProps {
    onLayout?: (event: LayoutChangeEvent) => void;
    onPress: () => void;
    testID?: string;
    disabled?: boolean;
    accessible?: boolean;
    rippleColor?: string;
    borderless?: boolean;
    style?: ViewStyle;
}

const Touchable: FunctionComponent<TouchableProps> = props => {
    if (isAndroid) {
        return (
            <TouchableNativeFeedback
                {...props}
                useForeground={TouchableNativeFeedback.canUseNativeForeground()}
                background={TouchableNativeFeedback.Ripple(
                    props.rippleColor ?? DEFAULT_RIPPLE_COLOR,
                    !!props.borderless
                )}
            >
                <View style={{ ...props.style, overflow: "hidden" }}>{props.children}</View>
            </TouchableNativeFeedback>
        );
    } else {
        if (props.underlayColor) {
            return (
                <TouchableHighlight {...props} underlayColor={props.underlayColor} activeOpacity={props.activeOpacity}>
                    <Fragment>{props.children}</Fragment>
                </TouchableHighlight>
            );
        }
        return <TouchableOpacity {...props}>{props.children}</TouchableOpacity>;
    }
};

export { Touchable };
