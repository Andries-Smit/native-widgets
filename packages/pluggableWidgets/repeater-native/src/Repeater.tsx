import { RepeaterProps } from "../typings/RepeaterProps";
import { defaultRepeaterStyle, RepeaterStyle } from "./ui/Styles";
import { createElement, ReactElement, useCallback, Fragment, useMemo } from "react";
import { View } from "react-native";
import { ActionValue, ValueStatus } from "mendix";
import { executeAction } from "@widgets-resources/piw-utils";
import { mergeNativeStyles } from "@mendix/pluggable-widgets-tools";
import { Touchable } from "./Touchable";

export function Repeater(props: RepeaterProps<RepeaterStyle>): ReactElement {
    const styles = mergeNativeStyles(defaultRepeaterStyle, props.style);
    if (
        props.datasource.status === ValueStatus.Loading ||
        !props.datasource.items ||
        props.datasource.items.length === 0
    ) {
        return <View />;
    }

    const onClick = useCallback((action: ActionValue) => {
        executeAction(action);
    }, []);

    const items = useMemo(
        () =>
            props.datasource.items?.map((item, index) => {
                const action = props.onClick?.(item);

                return action ? (
                    <Touchable key={index} onPress={() => onClick(action)} style={styles.container}>
                        {props.content(item)}
                    </Touchable>
                ) : (
                    <View key={index} style={styles.container}>
                        {props.content(item)}
                    </View>
                );
            }),
        [props.datasource, props.onClick, props.content]
    );

    return <Fragment>{items}</Fragment>;
}
