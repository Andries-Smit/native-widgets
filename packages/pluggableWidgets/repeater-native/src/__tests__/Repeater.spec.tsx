import { fireEvent, render } from "react-native-testing-library";
import { Repeater } from "../Repeater";
import { ListValueBuilder } from "@widgets-resources/piw-utils";
import { createElement } from "react";
import { Text } from "react-native";
import { Touchable } from "../Touchable";

describe("Repeater", () => {
    it("renders correctly", () => {
        const component = render(
            <Repeater
                name="test-repeater"
                style={[{ container: {} }]}
                datasource={ListValueBuilder().withAmountOfItems(5)}
                content={() => <Text>Item</Text>}
            />
        );

        expect(component.toJSON()).toMatchSnapshot();
    });

    it("renders correctly with custom styles", () => {
        const component = render(
            <Repeater
                name="test-repeater"
                style={[{ container: { backgroundColor: "red" } }]}
                datasource={ListValueBuilder().withAmountOfItems(5)}
                content={() => <Text>Item</Text>}
            />
        );

        expect(component.toJSON()).toMatchSnapshot();
    });

    it("renders correctly containing action", () => {
        const onClick = jest.fn();
        const component = render(
            <Repeater
                name="test-repeater"
                style={[{ container: {} }]}
                datasource={ListValueBuilder().withAmountOfItems(5)}
                content={() => <Text>Item</Text>}
                onClick={() => ({
                    canExecute: true,
                    isExecuting: false,
                    execute: onClick
                })}
            />
        );

        expect(component.toJSON()).toMatchSnapshot();
        fireEvent.press(component.getAllByType(Touchable).pop()!);
        expect(onClick).toHaveBeenCalledTimes(1);
    });

    it("renders correctly containing action and custom styles", () => {
        const component = render(
            <Repeater
                name="test-repeater"
                style={[{ container: { backgroundColor: "green" } }]}
                datasource={ListValueBuilder().withAmountOfItems(5)}
                content={() => <Text>Item</Text>}
                onClick={() => ({
                    canExecute: true,
                    isExecuting: false,
                    execute: jest.fn()
                })}
            />
        );

        expect(component.toJSON()).toMatchSnapshot();
    });
});
