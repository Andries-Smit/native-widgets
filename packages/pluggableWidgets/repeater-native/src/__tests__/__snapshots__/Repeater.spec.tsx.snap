// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repeater renders correctly 1`] = `
Array [
  <View
    style={Object {}}
  >
    <Text>
      Item
    </Text>
  </View>,
  <View
    style={Object {}}
  >
    <Text>
      Item
    </Text>
  </View>,
  <View
    style={Object {}}
  >
    <Text>
      Item
    </Text>
  </View>,
  <View
    style={Object {}}
  >
    <Text>
      Item
    </Text>
  </View>,
  <View
    style={Object {}}
  >
    <Text>
      Item
    </Text>
  </View>,
]
`;

exports[`Repeater renders correctly containing action 1`] = `
Array [
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text>
      Item
    </Text>
  </View>,
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text>
      Item
    </Text>
  </View>,
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text>
      Item
    </Text>
  </View>,
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text>
      Item
    </Text>
  </View>,
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text>
      Item
    </Text>
  </View>,
]
`;

exports[`Repeater renders correctly containing action and custom styles 1`] = `
Array [
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "green",
        "opacity": 1,
      }
    }
  >
    <Text>
      Item
    </Text>
  </View>,
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "green",
        "opacity": 1,
      }
    }
  >
    <Text>
      Item
    </Text>
  </View>,
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "green",
        "opacity": 1,
      }
    }
  >
    <Text>
      Item
    </Text>
  </View>,
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "green",
        "opacity": 1,
      }
    }
  >
    <Text>
      Item
    </Text>
  </View>,
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "green",
        "opacity": 1,
      }
    }
  >
    <Text>
      Item
    </Text>
  </View>,
]
`;

exports[`Repeater renders correctly with custom styles 1`] = `
Array [
  <View
    style={
      Object {
        "backgroundColor": "red",
      }
    }
  >
    <Text>
      Item
    </Text>
  </View>,
  <View
    style={
      Object {
        "backgroundColor": "red",
      }
    }
  >
    <Text>
      Item
    </Text>
  </View>,
  <View
    style={
      Object {
        "backgroundColor": "red",
      }
    }
  >
    <Text>
      Item
    </Text>
  </View>,
  <View
    style={
      Object {
        "backgroundColor": "red",
      }
    }
  >
    <Text>
      Item
    </Text>
  </View>,
  <View
    style={
      Object {
        "backgroundColor": "red",
      }
    }
  >
    <Text>
      Item
    </Text>
  </View>,
]
`;
